### PR TITLE
Next js prototype

### LINKED ISSUE #2898: ESlint support for typescript
**Is your feature request related to a problem? Please describe.**
More and more front-end is refactored towards typescript, but ESLint only runs on javascript files.

ESLint should also lint javascript in `.tsx` type files

### LINKED ISSUE #3793: Deploy nextjs server
# Description

We aim to use Next.js meta-framework. To utilize the full feature set Next.js offers we decided to run next.js server in production. This means we'll need to handle the server process, output and changes in the build.

# Tasks
- [x] Start/stop nextjs server with island startup/shutdown - @VakarisZ 
- [x] Configure nextjs logging to output to a file in the same directory as island logs - @VakarisZ 
- [x] Add configurable port - @VakarisZ 
- [x] Update the deployment scripts to install the node server @cakekoa 
- [ ] ~Duplicate the Island's jenkins jobs~
    - [ ] ~MSI~
    - [ ] ~AppImage~
    - [ ] ~Docker~
- [x] Edit duplicate jobs to use nextjs build commands and artifacts @cakekoa 
    - [x] MSI (need to update the MSI script in the build repo)
    - [x] AppImage (need to update build_scripts/common.sh in the monkey repo)
    - [x] Docker (same as appimage)
- [x] Smoke test the builds, make sure their size didn't increase dramatically
    - AppImage increased from 158 -> 199MB
    - Docker increased from 168 -> 208MB
- [x] Fix certificate configuration
- [x] Extend docs to explain where nextjs logs can be found
- [x] Extend the docs to explain how to configure the port
- [ ] Add ~development~ feature flag to not run the production during development OR configure dev script

### LINKED ISSUE #3794: Barebones UI prototype
# Description

Create a barebones UI prototype utilizing nextjs. This prototype will be a POC and will be used to alter build pipelines.

# Pre-requisite

If you take this task make sure to familiarize with #3678 

# Requirements

**Authentication** - use the same mechanism we currently use, don't use any libs. Fix up any obvious bugs if you want.
**Island communication** - UI needs to get a list of machines and be able to display a specific machine in a dynamic route. Page doesn't need too look good, the important part is to have a dynamic route that does real end-to-end communication.

Starting point for this feature should be code from [ui-prototypes branch](https://github.com/guardicore/monkey/tree/ui-prototypes). 

# Tasks

- [x] Implement registration - @VakarisZ 
- [x] Implement login - @VakarisZ 
- [x] Implement homepage that lists all machines - @VakarisZ 
- [x] Implement a dynamic route that provides information about specific machine based on its ID - @VakarisZ

### PR DESCRIPTION
**Pull request by VakarisZ**
# What does this PR do?

Fixes #3793 
Fixes #3794 
Fixes #2898 

PR adds next.js server and UI prototype using a feature flag.

## PR Checklist
* [ ] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [ ] Added relevant unit tests?
* [ ] Do all unit tests pass?
* [x] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by running with/without feature flag set
* [ ] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: build_scripts/common.sh [hunk 1]
```diff
@@ -73,30 +78,32 @@ generate_ssl_cert() {
...
   log_message "Generating front end"
   npm ci
   if [ "$is_release_build" == true ]; then
     log_message "Running production front end build"
-    npm run dist
+    npm run build
   else
     log_message "Running development front end build"
-    npm run dev
+    npm run build # Same as production build?
```
Comment (VakarisZ) [inline build_scripts/common.sh:92]: We need to use a feature flag here as well

File: build_scripts/appimage/AppRun [hunk 2]
```diff
@@ -24,7 +29,19 @@ do
     fi
 done
 
+
 export PYTHONNOUSERSITE=1
-(PYTHONHOME="${APPDIR}/opt/python3.11" exec "${APPDIR}/opt/python3.11/bin/python3.11" "${APPDIR}/usr/src/monkey_island.py" $@)
+
+export run_island="${APPDIR}/opt/python3.11/bin/python3.11 ${APPDIR}/usr/src/monkey_island.py"
```
Comment (ilija-lazoroski) [inline build_scripts/appimage/AppRun:35]: Environmental variables are usually all in caps. https://google.github.io/styleguide/shellguide.html#constants-and-environment-variable-names

File: build_scripts/common.sh [hunk 3]
```diff
@@ -8,10 +10,7 @@ copy_monkey_island_to_build_dir() {
   cp "$src"/monkey_island.py "$build_dir"
   cp -r "$src"/common "$build_dir/"
 
-  rsync \
-      -ar \
-      --exclude=monkey_island/cc/ui/node_modules \
-      --exclude=monkey_island/cc/ui/.npm \
```
Comment (ilija-lazoroski) [inline build_scripts/common.sh:14]: Why are we not excluding these anymore?

File: docs/content/setup/aws.md [hunk 4]
```diff
@@ -22,12 +22,17 @@ At this point, AWS will instance and deploy the new machine.
...
-`https://{public-ip}:5000`
+`https://{public-ip}`
 
 To login to the machine, use *ubuntu* username.
 
 Once you have access to the Monkey Island server, check out the [getting started page]({{< ref "/usage/getting-started" >}}).
 
+## Configuration
+
+Azure VM is running the AppImage deployment of the Infection Monkey. To configure the VM, shell
```
Comment (ilija-lazoroski) [inline docs/content/setup/aws.md:33]: `Azure VM` in AWS documentation?

File: monkey/monkey_island/cc/next_ui/src/providers/theme/EmotionCache.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        });
+
+        return (
+            <React.Fragment>
+                {globals.map(({ name, style }) => (
+                    <style
+                        key={name}
+                        data-emotion={`${registry.cache.key}-global ${name}`}
+                        // eslint-disable-next-line react/no-danger
+                        dangerouslySetInnerHTML={{ __html: style }}
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/next_ui/src/providers/theme/EmotionCache.tsx:84]: Is this temporary? We shouldn't use `dangerouslySetInnerHTML` because well, the name says it.

File: build_scripts/appimage/AppRun (hunk 2 above)
Comment (VakarisZ) [reply build_scripts/appimage/AppRun:35]: Should be addressed by https://github.com/guardicore/monkey/pull/3955

File: monkey/monkey_island/cc/next_ui/src/providers/theme/EmotionCache.tsx (hunk 5 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/next_ui/src/providers/theme/EmotionCache.tsx:84]: @ordabach do you know where the cache is coming from?

File: monkey/monkey_island/cc/next_ui/next.config.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    compiler: {
+        removeConsole: keepConsoleErrors
+    },
+    output: 'standalone',
+    typescript: {
+        // !! WARN !!
+        // Dangerously allow production builds to successfully complete even if
+        // your project has type errors.
+        // !! WARN !!
+        ignoreBuildErrors: true
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/next_ui/next.config.js:21]: Why do we allow this?

File: monkey/monkey_island/cc/next_ui/src/app/api/auth/[...nextauth]/authOptions.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+                } catch (error) {
+                    console.error('Error refreshing access token', error);
+
+                    return null; // clear the session if the refresh token gets an error
+
+                    // The error property will be used client-side to handle the refresh token error
+                    // return {
+                    //     ...token,
+                    //     error: 'RefreshAccessTokenError' as const
+                    // };
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/next_ui/src/app/api/auth/[...nextauth]/authOptions.ts:134]: Are we going to do this later?

File: monkey/monkey_island/cc/next_ui/src/app/api/user/signout/route.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                    method: HTTP_METHODS.POST,
+                    headers: {
+                        'Content-Type': 'application/json',
+                        'Authentication-Token': session.accessToken
+                    }
+                }
+            );
+
+            const logoutResponseBody = await logoutResponse.json();
+            console.log('logoutResponseBody', logoutResponseBody);
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/next_ui/src/app/api/user/signout/route.ts:23]: Do we need this?

File: monkey/monkey_island/cc/next_ui/src/app/consts.tsx [hunk 9]
```diff
@@ -0,0 +1,6 @@
+function getJSRuntimePort() {
+    // TODO remove with the NEXT_JS_UI_FEATURE in monkey_island/cc/feature_flags.py:1
+    return process.env.NEXT_PUBLIC_JAVASCRIPT_RUNTIME_PORT || 5000;
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/next_ui/src/app/consts.tsx:3]: Can we do this now?

Comment (VakarisZ) [reply monkey/monkey_island/cc/next_ui/src/app/consts.tsx:3]: Not yet, we'll do it once the old UI is removed

File: monkey/monkey_island/cc/next_ui/src/app/api/auth/[...nextauth]/authOptions.ts (hunk 7 above)
Comment (VakarisZ) [reply monkey/monkey_island/cc/next_ui/src/app/api/auth/[...nextauth]/authOptions.ts:134]: The whole authentication needs a lot of work

**Review by cakekoa — COMMENTED** [review]
A few comments. I didn't review everything since I needed to learn about Redux and NextAuth

File: CHANGELOG.md [hunk 10]
```diff
@@ -6,8 +6,8 @@ The format is based on [Keep a
 Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## [Unreleased]
-
 ### Added
+- Configurable Island port through node proxy server. #3827
```
[the remark this answers is not in the document]
Comment (VakarisZ) [reply CHANGELOG.md:10]: AFAIK `--network=host` already exposes the port.

File: deployment_scripts/deploy_linux.sh [hunk 11]
```diff
@@ -235,7 +239,7 @@ pushd "$ISLAND_PATH/cc/ui" || handle_error
 npm ci
 
 log_message "Generating front end"
-npm run dev
+npm run build
```
[the remark this answers is not in the document]
Comment (VakarisZ) [reply deployment_scripts/deploy_linux.sh:242]: I'm not sure what value the `dev` build had, but on the new UI we can build production or run a dev server with hot reloads. This builds production in case the dev is not intending to work on UI.

File: docs/content/setup/docker.md [hunk 12]
```diff
@@ -107,6 +107,23 @@ You can configure the server by mounting a volume and specifying a
...
+1. Stop the Monkey Island process.
+1. Modify the `server_config.json` by adding the following lines:
+    ```json
+    {
+      ...
+      "island_port": 8080,
+      ...
+    }
+    ```
+1. Run the Monkey Island again, it will be accessible on `https://localhost:8080`.
```
[the remark this answers is not in the document]
Comment (VakarisZ) [reply docs/content/setup/docker.md:125]: I don't think so, but I haven't tested it IIRC

File: docs/content/setup/linux.md [hunk 13]
```diff
@@ -27,6 +27,12 @@ On Windows, AppImage can be run in WSL 2.
 
 ## Deployment
 
+{{% notice info %}}
+You need to have **root permissions** to run the Infection Monkey AppImage package.
+By default, the Infection Monkey server is configured to listen on **port 443**,
+which requires setting the `CAP_NET_BIND_SERVICE` capability on the AppImage package.
```
[the remark this answers is not in the document]
Comment (VakarisZ) [reply docs/content/setup/linux.md:33]: What I was trying to say is that the process will not run as root, but root permissions are required. Feel free to rephrase or shorten

File: monkey/monkey_island/cc/next_ui/README.md
[the remark this answers is not in the document]
Comment (VakarisZ) [reply monkey/monkey_island/cc/next_ui/README.md:1]: Yeah, deleted

File: CHANGELOG.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (cakekoa) [reply CHANGELOG.md:10]: That may be correct. I tied to test out this feature in docker and couldn't access the UI. I realized that it's only applicable to the old UI but I still couldn't get it to work

File: deployment_scripts/deploy_linux.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (cakekoa) [reply deployment_scripts/deploy_linux.sh:242]: Let me rephrase. This is generating the frontend for the old UI (`"$ISLAND_PATH/cc/ui"`). Does `build` work for it?
